### PR TITLE
Remove Router

### PR DESCRIPTION
**Pull request by Ramarti**
Remove routing code in favour of more explicit interfaces.

### CODE REVIEW

File: contracts/components/utils/Routed.sol [hunk 1]
```diff
@@ -4,43 +4,22 @@
...
-    /**
-     * @notice Initializer method, access point to initialize inheritance tree.
-     * @param router address of Routed.
-     */
-    function __Routed_init(address router) internal initializer {
-        if (router == address(0)) revert ZeroAddress("router");
-        _router = IRouter(router);
-        emit RouterUpdated(router);
-    }
+    /// @custom:oz-renamed-from router
```
Comment (frangio) [inline contracts/components/utils/Routed.sol:14]:
```suggestion
    /// @custom:oz-renamed-from _router
```

File: contracts/components/utils/Routed.sol [hunk 2]
```diff
@@ -4,43 +4,22 @@
...
-    }
+    event RouterUpdated(address indexed router);
 
-    /// Sets new Router instance. Restricted to DEFAULT_ADMIN_ROLE.
-    function setRouter(address newRouter) public onlyRole(DEFAULT_ADMIN_ROLE) {
-        if (newRouter == address(0)) revert ZeroAddress("newRouter");
-        _router = IRouter(newRouter);
-        emit RouterUpdated(newRouter);
+    /// Sets Router instance to address(0). Restricted to DEFAULT_ADMIN_ROLE.
+    function disableRouter() public onlyRole(DEFAULT_ADMIN_ROLE) {
```
Comment (frangio) [inline contracts/components/utils/Routed.sol:20]: Why is this necessary? If for cleanup, can we do it in a reinitializer? (Related: #127)

File: contracts/components/scanners/ScannerRegistryEnable.sol [hunk 3]
```diff
@@ -132,7 +132,7 @@ abstract contract ScannerRegistryEnable is ScannerRegistryManaged {
      * @param value true if enabling, false if disabling.
      */
     function _afterScannerEnable(uint256 scannerId, Permission permission, bool value) internal virtual {
```
Comment (frangio) [inline contracts/components/scanners/ScannerRegistryEnable.sol:134]: Do we plan to remove these hooks in a future PR?

File: test/components/upgrades.test.js [hunk 4]
```diff
@@ -12,23 +12,25 @@ describe('Upgrades testing', function () {
...
             await agents.prepareAgent(prepareCommit(...args));
             await network.provider.send('evm_increaseTime', [300]);
-            await expect(agents.connect(this.accounts.other).createAgent(...args));
+            await agents.connect(this.accounts.user1).createAgent(...args);
 
             // Checks
-            //expect(await this.agents.isCreated(AGENT_ID)).to.be.equal(true); //Does not exist in 0.1.1
             expect(await agents.name()).to.be.equal('Forta Agents');
             expect(await agents.symbol()).to.be.equal('FAgents');
+            console.log('lol');
```
Comment (frangio) [inline test/components/upgrades.test.js:33]: A few leftover console.logs in this file. :stuck_out_tongue_winking_eye:

File: test/components/scannerversion.test.js [hunk 5]
```diff
@@ -48,7 +48,7 @@ describe('Scanner Node Software Version', function () {
     describe('upgrade', async function () {
         it('sets version', async function () {
             const ScannerVersion_0_1_0 = await ethers.getContractFactory('ScannerNodeVersion_0_1_0');
-            const originalScannerVersion = await upgrades.deployProxy(ScannerVersion_0_1_0, [this.contracts.access.address, this.contracts.router.address], {
+            const originalScannerVersion = await upgrades.deployProxy(ScannerVersion_0_1_0, [this.contracts.access.address, this.contracts.access.address], {
```
Comment (frangio) [inline test/components/scannerversion.test.js:51]: We shouldn't be using a proper contract in place of the router. I would prefer to use an empty contract as a mock router. Same applies to the other test files.

File: test/components/upgrades.test.js [hunk 6]
```diff
@@ -314,6 +327,123 @@ describe('Upgrades testing', function () {
             expect(await this.staking.activeSharesToStake(subjectToActive(0, this.accounts.scanner.address), '10')).to.be.equal('10');
             expect(await this.staking.inactiveSharesToStake(subjectToInactive(0, this.accounts.scanner.address), '10')).to.be.equal('10');
 
+            expect(await this.staking.stakeToActiveShares(subjectToActive(1, AGENT_ID), '10')).to.be.equal('20');
+            expect(await this.staking.stakeToInactiveShares(subjectToInactive(1, AGENT_ID), '10')).to.be.equal('10');
+            expect(await this.staking.activeSharesToStake(subjectToActive(1, AGENT_ID), '10')).to.be.equal('5');
+            expect(await this.staking.inactiveSharesToStake(subjectToInactive(1, AGENT_ID), '10')).to.be.equal('10');
+        });
+    });
+    describe('FortaStaking', async function () {
```
Comment (frangio) [inline test/components/upgrades.test.js:336]: Is this related to this PR?

File: contracts/components/scanners/ScannerRegistryEnable.sol (hunk 3 above)
Comment (Ramarti) [reply contracts/components/scanners/ScannerRegistryEnable.sol:134]: I thought about it several times, never came up. I'm doing the new `NodeRunnerRegistry` without them, so I'm all up for removing them from `AgentRegistry` too. I'll do a separate PR.

File: test/components/upgrades.test.js (hunk 4 above)
Comment (Ramarti) [reply test/components/upgrades.test.js:33]:
```
console.log(😅)
```
Are you using a script or tool to catch those?

File: contracts/components/_old/dispatch/Dispatch_0_1_4.sol [hunk 7]
```diff
@@ -0,0 +1,4319 @@
+// Sources flattened with hardhat v2.9.9 https://hardhat.org
```
[the remark this answers is not in the document]
Comment (Ramarti) [reply contracts/components/_old/dispatch/Dispatch_0_1_4.sol:1]: [The process is here](https://github.com/forta-network/forta-contracts#development-of-upgrades) but it should be expanded on the file generation step.

The files are generated following this process: 
1. hardhat flattener
2. Good ol' manual removing of repeated `pragma` statements, redundant licensing...
3. Renaming the top level file so it does not generates conflicts in the tests.

The purpose is testing the storage layout for the upgrades. The truth is this process is a PITA, I didn't fully script it because I have the muscle memory down by this point, did not want to go into forking the flattener and scripting text modification with regex usually ends up being a time sink.

This was needed before `upgrades-plugin` checked for renaming and `__gap` array modification between upgrades, but maybe now the process of upgrade testing is obsolete.

In any case I was thinking on just testing the last upgrade from now on, because every file adds to compilation and test execution time, and they are big and ugly.

What do you think?

File: test/components/upgrades.test.js (hunk 6 above)
Comment (Ramarti) [reply test/components/upgrades.test.js:336]: Not really but the tests were failing and I had to merge dev to `chore/remove_router` to reproduce the problem. That test came along

File: contracts/components/_old/dispatch/Dispatch_0_1_4.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (frangio) [reply contracts/components/_old/dispatch/Dispatch_0_1_4.sol:1]: >but maybe now the process of upgrade testing is obsolete

No, I think this is something we should continue to do. At least some kind of smoke test. But there is no need to keep the source code and keep compiling them all the time. We can just keep the JSON artifact.

File: test/components/upgrades.test.js (hunk 4 above)
Comment (frangio) [reply test/components/upgrades.test.js:33]: No just reviewing the code!

File: test/components/upgrades.test.js (hunk 6 above)
Comment (frangio) [reply test/components/upgrades.test.js:336]: Oh so this test is already in `dev`? Weird that it shows up in the diff.

File: test/components/upgrades.test.js (hunk 4 above)
Comment (Ramarti) [reply test/components/upgrades.test.js]: Removed in 6a30940

File: test/components/scannerversion.test.js (hunk 5 above)
Comment (Ramarti) [reply test/components/scannerversion.test.js:51]: Fair enough e2f1d36

File: contracts/components/utils/Routed.sol (hunk 2 above)
Comment (Ramarti) [reply contracts/components/utils/Routed.sol:20]: It is for cleanup.
We can and we should use reinitializers (now that you kindly reminded me that they are a thing), but in this version of OZ upgradeable, `_version` is not exposed and all of the components inherit from `Router`. 

I would not know how to call the `reinitializer` with a number that works for all of them, so I would let `disableRouter` stand in this PR, and continue the discussion in #127 . I'm open to upgrading OZ to 4.8 (if it's safe) before the code is audited to implement this properly.

Does that look good to you?

File: README.md [hunk 8]
```diff
@@ -132,15 +132,16 @@ The process to develop, test and deploy new implementations are as follow:
...
-    - Use `hardhat flattener` to generate one.
-    - If the contracts is deployed and the repo is not tagged and you are not sure if the exact files are in the current commit, generate a flattened version obtained from verified contracts in relevant block explorer (Etherscan, Polyscan...) and add it.
-    - Rename main component in flattened file to `<ComponentName>_X_Y_Z`.
-4.2 Initialize all the state of the contract, assert the values are there.
-4.3 Upgrade to the new implementation
-4.4 Check the output from `upgrades-plugin`. If gap sizes changed, you may need to activate the flag `unsafeSkipStorageCheck` to deploy.
-4.5 Assert the values of the state have not changed.
+    1 Deploy an upgradeable proxy with the base implementation, found in `components/_old/<component_folder>/<ComponentName>_X_Y_Z`. 
+        If not there, either:
+        - Use `hardhat flattener` to generate a file with all the inheriting contracts.
```
Comment (frangio) [inline README.md:137]:
```suggestion
        - Use `hardhat flatten` to generate a file with all the inheriting contracts.
```

File: README.md [hunk 9]
```diff
@@ -132,15 +132,16 @@ The process to develop, test and deploy new implementations are as follow:
...
+        - Remove duplicate licenses and pragma statements.
+        - Rename main component in flattened file to `<ComponentName>_X_Y_Z`.
+        - If the contracts is deployed and the repo is not tagged and you are not sure if the exact files are in the current commit, generate a flattened version obtained from verified contracts in relevant block explorer (Etherscan, Polyscan...) and add it.
+    2 Initialize all the state of the contract, assert the values are there.
+    3 Upgrade to the new implementation
+    4 Check the output from `upgrades-plugin`. If gap sizes changed, you may need to activate the flag `unsafeSkipStorageCheck` to deploy.
```
Comment (frangio) [inline README.md:143]: The part about `unsafeSkipStorageCheck` no longer applies right?

File: contracts/components/utils/Routed.sol (hunk 2 above)
Comment (frangio) [reply contracts/components/utils/Routed.sol:20]: >I would not know how to call the reinitializer with a number that works for all of them

This is a good point. Ok with leaving it like this but would slightly prefer removing `onlyRole` and reverting if the router is already zero.

File: test/components/upgrades.test.js [hunk 10]
```diff
@@ -199,15 +209,16 @@ describe('Upgrades testing', function () {
             const DELAY = 123123;
             const TREASURY = await ethers.Wallet.createRandom().address;
             const FortaStaking_0_1_0 = await ethers.getContractFactory('FortaStaking_0_1_0');
-            this.staking = await upgrades.deployProxy(FortaStaking_0_1_0, [this.access.address, this.router.address, this.token.address, DELAY, TREASURY], {
+            const routerMock = this.token.address;
+            this.staking = await upgrades.deployProxy(FortaStaking_0_1_0, [this.access.address, routerMock, this.token.address, DELAY, TREASURY], {
```
Comment (frangio) [inline test/components/upgrades.test.js:213]:
```suggestion
            this.staking = await upgrades.deployProxy(FortaStaking_0_1_0, [this.access.address, mockRouter, this.token.address, DELAY, TREASURY], {
```

File: test/components/upgrades.test.js [hunk 11]
```diff
@@ -199,15 +209,16 @@ describe('Upgrades testing', function () {
...
                 kind: 'uups',
                 constructorArgs: [this.contracts.forwarder.address],
                 unsafeAllow: ['delegatecall'],
             });
             await this.staking.deployed();
 
             const FortaStakingParameters_0_1_0 = await ethers.getContractFactory('FortaStakingParameters_0_1_0');
-            this.stakingParameters = await upgrades.deployProxy(FortaStakingParameters_0_1_0, [this.access.address, this.router.address, this.staking.address], {
+            this.stakingParameters = await upgrades.deployProxy(FortaStakingParameters_0_1_0, [this.access.address, routerMock, this.staking.address], {
```
Comment (frangio) [inline test/components/upgrades.test.js:221]:
```suggestion
            this.stakingParameters = await upgrades.deployProxy(FortaStakingParameters_0_1_0, [this.access.address, mockRouter, this.staking.address], {
```

File: test/components/upgrades.test.js [hunk 12]
```diff
@@ -314,6 +325,123 @@ describe('Upgrades testing', function () {
...
+    describe('FortaStaking', async function () {
+        it('0.1.0 -> 0.1.1', async function () {
+            this.accounts.getAccount('scanner');
+            const STAKING_PARAMS = { max: '1000', min: '100', activated: true };
+
+            const DELAY = 123123;
+            const TREASURY = await ethers.Wallet.createRandom().address;
+            const FortaStaking_0_1_0 = await ethers.getContractFactory('FortaStaking_0_1_0');
+            const fakeRouter = ethers.Wallet.createRandom().address;
+            this.staking = await upgrades.deployProxy(FortaStaking_0_1_0, [this.access.address, fakeRouter, this.token.address, DELAY, TREASURY], {
```
Comment (frangio) [inline test/components/upgrades.test.js:343]:
```suggestion
            this.staking = await upgrades.deployProxy(FortaStaking_0_1_0, [this.access.address, mockRouter, this.token.address, DELAY, TREASURY], {
```

File: test/components/upgrades.test.js [hunk 13]
```diff
@@ -314,6 +325,123 @@ describe('Upgrades testing', function () {
...
+                kind: 'uups',
+                constructorArgs: [this.contracts.forwarder.address],
+                unsafeAllow: ['delegatecall'],
+            });
+            await this.staking.deployed();
+
+            const FortaStakingParameters_0_1_0 = await ethers.getContractFactory('FortaStakingParameters_0_1_0');
+            this.stakingParameters = await upgrades.deployProxy(FortaStakingParameters_0_1_0, [this.access.address, fakeRouter, this.staking.address], {
```
Comment (frangio) [inline test/components/upgrades.test.js:351]:
```suggestion
            this.stakingParameters = await upgrades.deployProxy(FortaStakingParameters_0_1_0, [this.access.address, mockRouter, this.staking.address], {
```

File: contracts/components/utils/Routed.sol (hunk 2 above)
Comment (Ramarti) [reply contracts/components/utils/Routed.sol]: 5fed729

File: README.md (hunk 9 above)
Comment (Ramarti) [reply README.md]: True! removed it from tests and scripts too aa0cc3e
